program-runtime/src/loaded_programs.rs

```diff
@@ -317,14 +321,44 @@ impl LoadedPrograms {
     }
 
     /// Evicts programs which were used infrequently
-    pub fn sort_and_evict(&mut self) {
-        // TODO: Sort programs by their usage_counter
-        // TODO: Truncate the end of the list
+    pub fn sort_and_evict(&mut self, max_cache_entries: Option<usize>) -> Vec<Pubkey> {
+        // Find eviction candidates and sort by their usage counters
+        let mut num_cache_entries = 0;
+        let sorted_candidates = self.entries
+            .iter_mut()
+            .filter(|(_key, programs)| {
+                num_cache_entries += programs.len();
+                programs.len() == 1
+            })
+            .sorted_by_cached_key(|(_key, programs)| programs[0].usage_counter.load(Ordering::Relaxed))
+            .map(|(key, _programs)| *key)
+            .collect::<Vec<Pubkey>>();
+        // Calculate how many to remove
+        let num_to_remove = std::cmp::min(
+            num_cache_entries.saturating_sub(max_cache_entries.unwrap_or(MAX_CACHE_ENTRIES)), 
+            sorted_candidates.len()
+        );
+        // Remove selected entries
+        if num_to_remove != 0 {
+            self.remove_entries(
+                sorted_candidates
+                    .into_iter()
+                    .take(num_to_remove)
+            )
+        }
+        else {
+            Vec::<Pubkey>::new()
+        }
     }
 
     /// Removes the entries at the given keys, if they exist
-    pub fn remove_entries(&mut self, _key: impl Iterator<Item = Pubkey>) {
-        // TODO: Remove at primary index level
+    pub fn remove_entries(&mut self, keys: impl Iterator<Item = Pubkey>) -> Vec<Pubkey> {
+        let mut removed_entires = Vec::<Pubkey>::new();
+        for k in keys {
+            self.entries.remove(&k);
+            removed_entires.push(k);
```

Comment:
would there ever be a case where `removed_entries` at return will not have all `keys` passed in as the parameter? If not, this is extra work we are doing building the vector of pubkeys.

Reply:
removed